### DISCUSSION TITLE
Wix v4 - HarvestDirectory + Windows Service Install

### DISCUSSION THREAD

**Discussion by vdschuck** [wixtoolset/issues#7461] [Questions] [answered]
I trying to build a project that runs heat and create a windows service, but I have duplicate component issue for example.exe file. How I should handle it?

_WIX0369: Component/@Id='WinServiceFileExe' with source path '...' has a @Guid value '...' that duplicates another component in this package. It is recommended to give each component its own unique GUID._ 

Heat example:

```
<ItemGroup>
        <HarvestDirectory Include=".\obj\Release\publish\">
            <ComponentGroupName>HarvestedComponents</ComponentGroupName>
            <DirectoryRefId>InstallFolder</DirectoryRefId>
            <SuppressRootDirectory>true</SuppressRootDirectory>
        </HarvestDirectory>
        <BindPath Include=".\obj\Release\publish\" />
</ItemGroup>
``` 
  
  
 Windows service component example:
  
```
<ComponentGroup Id="WindowsServiceComponents" Directory="InstallFolder">
            <Component>
                <File Id="WinServiceFileExe"
                    Name="example.exe"
                    KeyPath="yes"
                    Source="example.exe"/>
                <ServiceInstall Id="WinServiceInstaller"
                    Type="ownProcess"
                    Vital="yes"
                    Name="!(loc.ServiceName)"
                    DisplayName="!(loc.ServiceName)"
                    Description="!(loc.Description)"
                    ErrorControl="normal"
                    Account="LocalSystem"
                    Start="auto"
                    Interactive="no" />
                <ServiceControl Id="WinServiceControl"
                    Start="install"
                    Stop="both"
                    Remove="uninstall"
                    Name="!(loc.ServiceName)"
                    Wait="yes" />
            </Component>
</ComponentGroup>
```

**Comment by BMurri** [comment]
XSLT. You need to either remove the harvesting of `example.exe` or you need to use the transform to add the `ServiceInstall` and `ServiceControl` elements to the harvest-generated component for `example.com` and remove  your `WindowsServiceComponents`.

**Reply by vdschuck** [reply]
How could I remove `example.exe` from harvest? I can't do it manually, because the installer is going to be generate by CI.

What do you mean **_use the transform_**?

**Reply by BMurri** [reply, marked as answer]
XSLT seems to be viewed as an even darker magic art than deployment engineering is, but I personally have never understood the lack of appeal. Everyone who uses XML professionally should learn it, because it is the XML equivalent of regular expressions, and it is, quite frankly, easier to read than regular expressions (which to this day I have to consult one or more references every single time just to get a gist of what it's attempting to do).

Every harvest-related item has a `Transforms` metadata that you would point to one or more XSLT files. Heat would then run those transforms on the generated XML before writing out the generated `wxs` files that are added to the `Compile` msbuild item.

What you want to do is to start with the [identity transform](https://en.wikipedia.org/wiki/Identity_transform). Note that the XML parser that WiX uses only supports XSLT v1, so you can't use v3 syntax. You'll need to add all xml namespaces you will be using (at minimum that will include `xmlns:wix="http://wixtoolset.org/schemas/v4/wxs"`). Next, you will need to look at the output WXS file(s) from the harvester to see what is unique about the `Component` element you need to alter in order to build out your `template/@match` values.

To help you get started, here's the transform I used recently for harvesting a managed bootstrapper built using net6. In it, I remove one `Payload` element and add an attribute to another `Payload` element.

```
<?xml version="1.0" encoding="utf-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:msxsl="urn:schemas-microsoft-com:xslt" exclude-result-prefixes="msxsl"
    xmlns:wix="http://wixtoolset.org/schemas/v4/wxs" xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal">
  <xsl:output method="xml" indent="yes"/>

  <xsl:template match="/wix:Wix/wix:Fragment/wix:PayloadGroup/wix:Payload[@SourceFile='SourceDir\WpfBA.dll']">
    <xsl:copy>
      <xsl:apply-templates select="@*"/>
      <xsl:attribute name="bal:BAFactoryAssembly">yes</xsl:attribute>
      <xsl:apply-templates select="node()"/>
    </xsl:copy>
  </xsl:template>

  <xsl:template match="/wix:Wix/wix:Fragment/wix:PayloadGroup/wix:Payload[@SourceFile='SourceDir\WpfBA.exe']"/>

  <xsl:template match="@* | node()">
    <xsl:copy>
      <xsl:apply-templates select="@* | node()"/>
    </xsl:copy>
  </xsl:template>
</xsl:stylesheet>
```

**Reply by barnson** [reply]
XSLT is overkill. Create a staging directory for the things you want to harvest and omit the service .exe, which gets its own authoring.

**Reply by BMurri** [reply]
Or, recognize the fact that best practice is to not harvest during builds.

**Reply by BMurri** [reply]
To clarify my last comment: harvesting was created as an aid to generate the first version of your installer code, with the idea that it would be maintained in the fashion that @robmen demonstrates in his Deployment Dojo video series on YouTube (as well as explained on his various blog posts over the past couple of decades or so). The use of it during builds is never encouraged, but does facilitate use cases where maintaining those portions of the installer code directly related to the build outputs feeding the installer are subject to lots of changes across very disparate teams.

**Reply by barnson** [reply]
I'm fine with harvesting "content" -- but once a file requires special handling (like service authoring), Heat falls down. FireGiant's Harvesting does much better there. Staging directories let you harvest in bulk.

**Reply by robmen** [reply]
That's definitely true for `heat`. Heat was not designed to be run as part of the build. It's been shoehorned into that role, but there are many sharp edges (XSLT is one).

At FireGiant, we designed [Advanced Harvesting](https://www.firegiant.com/wix/wep-documentation/harvesting/) to be included in the build process. We spent a good amount of time designing it to work well. WiX v4 support is coming _very, very soon_.

**Comment by ghost** [comment]
I have the same problem like this, please find in the URL below the solution.

[https://github.com/z3ut/wix-toolset-net-core-service-example/tree/master](url)

You must add compiler tools to remove ICE30 validation

![image](https://github.com/wixtoolset/issues/assets/34740986/9260b8a3-a9f1-467a-8721-66a86b24271d)
